frame/identity/src/types.rs

```diff
@@ -0,0 +1,342 @@
+impl Encode for IdentityFields {
+	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+		self.0.bits().using_encoded(f)
+	}
+}
+impl Decode for IdentityFields {
+	fn decode<I: codec::Input>(input: &mut I) -> sp_std::result::Result<Self, codec::Error> {
+		let field = u64::decode(input)?;
+		Ok(Self(<BitFlags<IdentityField>>::from_bits(field as u64).map_err(|_| "invalid value")?))
+	}
+}
+
+/// Information concerning the identity of the controller of an account.
+///
+/// NOTE: This should be stored at the end of the storage item to facilitate the addition of extra
+/// fields in a backwards compatible way through a specialized `Decode` impl.
+#[derive(CloneNoBound, Encode, Decode, Eq, PartialEqNoBound, RuntimeDebugNoBound)]
+#[cfg_attr(test, derive(frame_support::DefaultNoBound))]
+pub struct IdentityInfo<FieldLimit: Get<u32>> {
+	/// Additional fields of the identity that are not catered for with the struct's explicit
+	/// fields.
+	pub additional: BoundedVec<(Data, Data), FieldLimit>,
+
+	/// A reasonable display name for the controller of the account. This should be whatever it is
+	/// that it is typically known as and should not be confusable with other entities, given
+	/// reasonable context.
+	///
+	/// Stored as UTF-8.
+	pub display: Data,
+
+	/// The full legal name in the local jurisdiction of the entity. This might be a bit
+	/// long-winded.
+	///
+	/// Stored as UTF-8.
+	pub legal: Data,
+
+	/// A representative website held by the controller of the account.
+	///
+	/// NOTE: `https://` is automatically prepended.
+	///
+	/// Stored as UTF-8.
+	pub web: Data,
+
+	/// The Riot/Matrix handle held by the controller of the account.
+	///
+	/// Stored as UTF-8.
+	pub riot: Data,
+
+	/// The email address of the controller of the account.
+	///
+	/// Stored as UTF-8.
+	pub email: Data,
+
+	/// The PGP/GPG public key of the controller of the account.
+	pub pgp_fingerprint: Option<[u8; 20]>,
+
+	/// A graphic image representing the controller of the account. Should be a company,
+	/// organization or project logo or a headshot in the case of a human.
+	pub image: Data,
+
+	/// The Twitter identity. The leading `@` character may be elided.
+	pub twitter: Data,
+}
+
+impl<FieldLimit: Get<u32>> MaxEncodedLen for IdentityInfo<FieldLimit> {
+	fn max_encoded_len() -> usize {
+		let mut len = 0usize;
+		len = len.saturating_add(<BoundedVec<(Data, Data), FieldLimit>>::max_encoded_len());
+		len = len.saturating_add(Data::max_encoded_len()); // display
+		len = len.saturating_add(Data::max_encoded_len()); // legal
+		len = len.saturating_add(Data::max_encoded_len()); // web
+		len = len.saturating_add(Data::max_encoded_len()); // riot
+		len = len.saturating_add(Data::max_encoded_len()); // email
+		len = len.saturating_add(<Option<[u8; 20]>>::max_encoded_len()); // PGP fingerprint
+		len = len.saturating_add(Data::max_encoded_len()); // image
+		len = len.saturating_add(Data::max_encoded_len()); // twitter
+		len
+	}
+}
```

Comment:
So we are giving up on the macros?

Reply:
See #9180. I would want to make a custom trait bound for it, but it'll have to wait until `MaxEncodedLen` is properly integrated with parity SCALE codec.

Reply:
Looks like https://github.com/paritytech/parity-scale-codec/pull/268 is merged, the next step is probably to wait for #9163 and me writing the code for parity SCALE codec to support custom trait bounds.

Reply:
Now waiting on paritytech/parity-scale-codec#279 for the derive.